File: src/MagnusAddon/note/vocabulary/related_vocab/related_vocab_data_serializer.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from note.notefields.auto_save_wrappers.value_wrapper import ValueWrapper
from note.notefields.json_object_field import ObjectSerializer
from sysutils.json import ex_json
from sysutils.json.json_reader import JsonReader

if TYPE_CHECKING:
    from note.vocabulary.related_vocab.related_vocab_data import RelatedVocabData

class RelatedVocabDataSerializer(ObjectSerializer["RelatedVocabData"], Slots):
    @override
    def deserialize(self, serialized: str) -> RelatedVocabData:
        from note.vocabulary.related_vocab.related_vocab_data import RelatedVocabData
        if not serialized: return RelatedVocabData("", ValueWrapper(""), set(), set(), set(), set(), set())

        reader = JsonReader.from_json(serialized)
        return RelatedVocabData(reader.string("ergative_twin"),
                                ValueWrapper(reader.string("derived_from")),
                                reader.string_set("perfect_synonyms", default=set()),
                                reader.string_set("synonyms"),
                                reader.string_set("antonyms"),
                                reader.string_set("confused_with"),
                                reader.string_set("see_also"))

    @override
    def serialize(self, instance: RelatedVocabData) -> str:
        return ex_json.dict_to_json({"ergative_twin": instance.ergative_twin,
                                     "derived_from": instance.derived_from.get(),
                                     "synonyms": list(instance.synonyms),
                                     "perfect_synonyms": list(instance.perfect_synonyms),
                                     "antonyms": list(instance.antonyms),
                                     "confused_with": list(instance.confused_with),
                                     "see_also": list(instance.see_also)})

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/RelatedVocab/RelatedVocabDataSerializer.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using JAStudio.Core.Note.NoteFields;
using JAStudio.Core.Note.NoteFields.AutoSaveWrappers;
using JAStudio.Core.SysUtils.Json;

namespace JAStudio.Core.Note.Vocabulary.RelatedVocab;

class RelatedVocabDataSerializer : IObjectSerializer<RelatedVocabData>
{
   readonly string _emptyObjectJson;

   public RelatedVocabDataSerializer() => _emptyObjectJson = Serialize(Deserialize(string.Empty));

   public RelatedVocabData Deserialize(string serialized)
   {
      if(string.IsNullOrEmpty(serialized))
      {
         return new RelatedVocabData(
            string.Empty,
            new ValueWrapper<string>(string.Empty),
            [],
            [],
            [],
            [],
            []
         );
      }

      using var doc = JsonDocument.Parse(serialized);
      var reader = new JsonReader(doc.RootElement);

      return new RelatedVocabData(
         reader.GetString("ergative_twin", string.Empty),
         new ValueWrapper<string>(reader.GetString("derived_from", string.Empty)),
         reader.GetStringSet("perfect_synonyms", []),
         reader.GetStringSet("synonyms", []),
         reader.GetStringSet("antonyms", []),
         reader.GetStringSet("confused_with", []),
         reader.GetStringSet("see_also", [])
      );
   }

   public string Serialize(RelatedVocabData instance)
   {
      var dict = new Dictionary<string, object>();

      if(!string.IsNullOrEmpty(instance.ErgativeTwin))
         dict["ergative_twin"] = instance.ErgativeTwin;

      if(!string.IsNullOrEmpty(instance.DerivedFrom.Get()))
         dict["derived_from"] = instance.DerivedFrom.Get();

      if(instance.Synonyms.Any())
         dict["synonyms"] = instance.Synonyms.OrderBy(s => s).ToList();

      if(instance.PerfectSynonyms.Any())
         dict["perfect_synonyms"] = instance.PerfectSynonyms.OrderBy(s => s).ToList();

      if(instance.Antonyms.Any())
         dict["antonyms"] = instance.Antonyms.OrderBy(s => s).ToList();

      if(instance.ConfusedWith.Any())
         dict["confused_with"] = instance.ConfusedWith.OrderBy(s => s).ToList();

      if(instance.SeeAlso.Any())
         dict["see_also"] = instance.SeeAlso.OrderBy(s => s).ToList();

      var json = JsonHelper.DictToJson(dict);
      return json != _emptyObjectJson ? json : string.Empty;
   }
}
